src/entity.ts

```diff
@@ -130,86 +128,119 @@ export class Entity {
-						state.forEach((s, idx) => {
-							if (!(s instanceof ChildEntity))
-								s = this.serializer.deserialize(this, s, prop, null, false);
-
-							// Modifying/replacing existing list item
-							if (idx < currentValue.length) {
-								// If the item is an object that has an Id property, then retrieve or create an object with that Id
-								if (!(s instanceof ChildEntity) && typeof s === "object" && getIdFromState(ChildEntity.meta, s))
-									s = ChildEntity.meta.createSync(s);
-								if (s instanceof ChildEntity)
-									state.splice(idx, 1, s);
-								else
-									currentValue[idx].set(s);
-							}
-							// Add a list item
-							else if (s instanceof ChildEntity)
-								currentValue.push(s);
-							else
-								currentValue.push(ChildEntity.meta.createSync(s));
-						});
+				if (this._context) {
+					const valueResolution = this._context.tryResolveValue(this, prop, state);
+					if (valueResolution) {
+						valueResolution.then(asyncState => this.setProp(prop, asyncState));
+						continue;
 					}
-					else if (state instanceof ChildEntity)
-						value = state;
-					else if (state == null)
-						value = null;
-					else {
-						// Attempt to deserialize the state
-						let newState = this.serializer.deserialize(this, state, prop, null, false);
-						if (typeof newState !== "undefined")
-							state = newState;
-						// Got null, so assign null to the property
-						if (state == null)
-							value = null;
-						// Got a valid instance, so use it
-						else if (state instanceof ChildEntity)
-							value = state;
-						// Got something other than an object, so just use it and expect to get a down-stream error
-						else if (typeof state !== "object")
-							value = state;
-						// Got an object, so attempt to fetch or create and assign the state
-						else if (getIdFromState(ChildEntity.meta, state))
-							value = ChildEntity.meta.createSync(state);
-						else if (currentValue)
-							currentValue.set(state);
+				}
+				this.setProp(prop, state);
+			}
+		}
+	}
+
+	private setProp(prop: Property, state: any) {
+		let value;
+		const currentValue = prop.value(this);
+		if (isEntityType(prop.propertyType)) {
+			const ChildEntity = prop.propertyType;
+			if (prop.isList && Array.isArray(state) && Array.isArray(currentValue)) {
+				state.forEach((s, idx) => {
+					if (!(s instanceof ChildEntity))
+						s = this.serializer.deserialize(this, s, prop, this._context, false);
+
+					// Modifying/replacing existing list item
+					if (idx < currentValue.length) {
+						// If the item is an object that has an Id property, then retrieve or create an object with that Id
+						if (!(s instanceof ChildEntity) && typeof s === "object" && getIdFromState(ChildEntity.meta, s)) {
+							const id = getIdFromState(ChildEntity.meta, s);
+							if (id && ChildEntity.meta.get(id))
+								ChildEntity.meta.get(id).withContext(this._context, entity => entity.set(state));
```

Comment:
If we don't assign `s` here, does that result in `set` being called twice?

Reply:
Fixed.